### PR TITLE
feat: TNLT-6350 Register Slice transform and Lead two no pic and two variant for sport section

### PR DESCRIPTION
**Pull request by justadamhopkins**
This PR is a combination of the slice work myself and @aronshamash have been working on. See corresponding tickets below.

[TNLT-6350-L2NP2 VARIANT](https://nidigitalsolutions.jira.com/browse/TNLT-6350)
[TNLT-6351 Register slice transform](https://nidigitalsolutions.jira.com/browse/TNLT-6351)

Changes made on this pr
1. Updated transform slice logic to consume a base config and merge overrides if they exist
2. Built out a new variant slice for lead two no pic and two which is built fully using configured tiles :-)
3. Built two configured tiles a standard column and column with image bottom. (These two should cover a lot of tiles)
4. Added transform for register slice
5. Added logic to in slice map to render the variant for the sport section only
6. Added and fixed a lot of tests

The main area to focus on is `transformSlice.ts`, `configuredTile` directory which currently contains two flexible tiles that are config driven and slice map switch for sport in the edition-slices` file

## Lead two no pic and two variant screenshots

Screenshot 2021-01-25 at 18 11 28
Screenshot 2021-01-25 at 17 36 17
Screenshot 2021-01-25 at 17 35 52

Screenshot 2021-01-25 at 18 11 48

Screenshot 2021-01-25 at 18 19 17

Screenshot 2021-01-25 at 18 22 07
 

# Lead One And One (Default)

### Landscape
Screenshot 2021-01-25 at 12 23 57

### Portrait
Screenshot 2021-01-25 at 12 24 14

### Landscape
Screenshot 2021-01-25 at 12 24 24

# Lead One And One (News section)

### Portrait
Screenshot 2021-01-25 at 12 24 42

### Landscape
Screenshot 2021-01-25 at 12 24 57

### Portrait

Screenshot 2021-01-25 at 12 27 05

### Landscape

Screenshot 2021-01-25 at 12 27 13

# Lead One And One (Register section)

### Portrait

Screenshot 2021-01-25 at 12 30 01

### Landscape

Screenshot 2021-01-25 at 12 30 13

### Portrait

Screenshot 2021-01-25 at 12 36 20

### Landscape

Screenshot 2021-01-25 at 12 36 20

### CODE REVIEW

File: packages/section/src/utils/transformSlice.ts [hunk 1]
```diff
@@ -1,30 +1,94 @@
...
+  },
 };
 
-const sliceTransformations: TransformSlice[] = [leadOneAndOneNewsTransform];
+const sliceTransformations: TransformSlice[] = [
+  leadOneAndOneNewsTransform,
+  leadOneAndOneRegisterTransform,
+];
 
 export const transformSlice = (isTablet: boolean, sectionTitle: string) => (
```
Comment (aronshamash) [inline packages/section/src/utils/transformSlice.ts:93]: for future PRs I think we should:
-  pass in the `sliceTransformations` to `transformSlice`
- sliceTransforms in separate file
- baseConfigs imported in separate file

File: packages/edition-slices/__tests__/ios/tile-col-with-Image-bottom.test.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import shared from "../tile-col-with-image-bottom/shared-tile-col-with-Image-bottom";
```
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/edition-slices/__tests__/ios/tile-col-with-Image-bottom.test.js:1]: I would call tile `tile-col-image-bottom`

File: packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  const config = tile.config[breakpoint];
+
+  const styles = styleFactory(config as TileConfig);
+
+  const renderTileImage = (
+    { article }: Tile,
+    imageProps: Pick<TileConfig, "image" | "portrait">,
+  ) => {
+    if (
+      imageProps.image?.orientation === orientation ||
```
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx:54]: yes exactly not props

File: packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+  const config = tile.config[breakpoint];
+
+  const styles = styleFactory(config as TileConfig);
+
+  const renderTileImage = (
+    { article }: Tile,
+    imageProps: Pick<TileConfig, "image" | "portrait">,
+  ) => {
+    if (
```
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx:53]: I'll update this logic

File: packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx:54]: I'll update

File: packages/edition-slices/src/configured-tiles/tile-col-standard/styles.tsx [hunk 5]
```diff
@@ -23,8 +17,8 @@ export default (breakpoint: string) => ({
   headline: {
     ...globalSpacingStyles.tabletHeadline,
     fontFamily: fonts.headline,
-    fontSize: fontSizeResolver[breakpoint],
-    lineHeight: fontSizeResolver[breakpoint],
+    fontSize: config?.headline?.fontSize,
```
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/edition-slices/src/configured-tiles/tile-col-standard/styles.tsx:20]: I'll have a look at it

File: packages/edition-slices/src/slices/leadtwonopicandtwovariant2/config.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  headline: { fontSize: 24 },
+  image: { ratio: "3:2", orientation: "landscape" },
+};
+
+const support2Base = {
+  image: { ratio: "2:3", orientation: "portrait" },
+  headline: { fontSize: 28 },
+};
+
+export const l2NoPic2V2Config: Omit<Slice, "name"> = {
```
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/edition-slices/src/slices/leadtwonopicandtwovariant2/config.ts:24]: I'll take a look

File: packages/section/src/section.tsx [hunk 7]
```diff
@@ -82,7 +88,7 @@ const Section: React.FC<Props> = ({
     return null;
   };
 
-  const renderItem = (isPuzzle: boolean) => ({
+  const renderItem = (isPuzzle: boolean, sectionTitle: SectionTitles) => ({
```
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/section/src/section.tsx]: I'll take a look

File: packages/edition-slices/__tests__/ios/tile-col-with-Image-bottom.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/edition-slices/__tests__/ios/tile-col-with-Image-bottom.test.js]: done

File: packages/edition-slices/src/configured-tiles/tile-col-standard/styles.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/edition-slices/src/configured-tiles/tile-col-standard/styles.tsx:20]: was able to remove null check

File: packages/edition-slices/src/slices/leadtwonopicandtwovariant2/config.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/edition-slices/src/slices/leadtwonopicandtwovariant2/config.ts:24]: updated

File: packages/section/__tests__/unit/transformSlice.test.ts [hunk 8]
```diff
@@ -1,16 +1,34 @@
...
+        config: {
+          mock: "mockSupport",
+        },
+      },
+    },
+  }),
+);
 
+describe("transformSlice", () => {
   describe("for the News section", () => {
```
[the remark this answers is not in the document]
Comment (justadamhopkins) [reply packages/section/__tests__/unit/transformSlice.test.ts]: This has been fixed!!

File: packages/edition-slices/src/slices/leadtwonopicandtwovariant2/config.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (justadamhopkins) [reply packages/edition-slices/src/slices/leadtwonopicandtwovariant2/config.ts]: This has been fixed

File: packages/edition-slices/src/configured-tiles/tile-col-image-bottom/index.tsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+      !imageProps.image?.orientation
+    ) {
+      const imageRatio =
+        imageProps.portrait && orientation === "portrait"
+          ? imageProps.portrait.ratio
+          : imageProps.image?.ratio;
+
+      if (!imageRatio) return null;
+
+      const crop: Crop = getTileImage(tile, getCropByRatio(imageRatio));
```
Comment (c-j-j) [inline packages/edition-slices/src/configured-tiles/tile-col-image-bottom/index.tsx:64]: getTileImage can return null, so this probably should be be `const crop: Crop | null = etc`

File: packages/edition-slices/src/edition-slices.js [hunk 10]
```diff
@@ -50,7 +54,10 @@ const sliceMap = (isInSupplement) => {
       ? SupplementLeadOneAndOneSlice
       : LeadOneAndOneSlice,
     LeadOneFullWidthSlice,
-    LeadTwoNoPicAndTwoSlice,
+    LeadTwoNoPicAndTwoSlice: renderLeadTwoNoPicAndTwoSlice(
```
[the remark this answers is not in the document]
Comment (c-j-j) [reply packages/edition-slices/src/edition-slices.js:57]: Any thoughts on this comment?

File: packages/fixture-generator/src/types.ts [hunk 11]
```diff
@@ -1481,7 +1481,7 @@ export interface EmbeddedComponentResource {
 
 /** An article presentation */
 export interface Tile {
-  articleId: Uuid;
+  articleId?: Uuid;
```
Comment (c-j-j) [inline packages/fixture-generator/src/types.ts:1484]: has this been manually changed - or updated via the schema scripts?

File: packages/section/src/section.tsx [hunk 12]
```diff
@@ -156,7 +168,7 @@ const Section: React.FC<Props> = ({
         nestedScrollEnabled
         onViewableItemsChanged={onViewed ? onViewableItemsChanged : null}
         {...(isPuzzle && { onScrollBeginDrag })}
-        renderItem={renderItem(isPuzzle)}
+        renderItem={renderItem(isPuzzle, orientation)}
```
Comment (c-j-j) [inline packages/section/src/section.tsx:171]: Not for this PR, but we're pulling `orientation` out of context and passing it as a prop to a child component quite a lot... whereas we could allow the child component to pull that out of context directly.

File: packages/edition-slices/__tests__/android/__snapshots__/tablet-slices-with-style.test.js.snap [hunk 13]
```diff
@@ -188,9 +188,7 @@ exports[`3. lead one and one - medium 1`] = `
           "width": "75%",
         }
       }
-    >
-      <TileU />
-    </View>
+    />
```
[the remark this answers is not in the document]
Comment (justadamhopkins) [reply packages/edition-slices/__tests__/android/__snapshots__/tablet-slices-with-style.test.js.snap]: This should be fixed now !!

File: packages/section/src/section.tsx (hunk 12 above)
Comment (aronshamash) [reply packages/section/src/section.tsx:171]: yes agreed we should do it as low down as we can.

File: packages/fixture-generator/src/types.ts (hunk 11 above)
Comment (aronshamash) [reply packages/fixture-generator/src/types.ts:1484]: I think this needs reverting as it was a manual change. You weren't to know but this file is generated from the TPA schema whenever someone runs a script.

Comment (aronshamash) [reply packages/fixture-generator/src/types.ts]: reverted

**Review by c-j-j — APPROVED** [review]

File: packages/edition-slices/src/configured-tiles/tile-col-standard/styles.tsx [hunk 14]
```diff
@@ -23,8 +17,8 @@ export default (breakpoint: string) => ({
   headline: {
     ...globalSpacingStyles.tabletHeadline,
     fontFamily: fonts.headline,
-    fontSize: fontSizeResolver[breakpoint],
-    lineHeight: fontSizeResolver[breakpoint],
+    fontSize: config.headline?.fontSize,
```
Comment (kamiranoff) [inline packages/edition-slices/src/configured-tiles/tile-col-standard/styles.tsx:20]: should this have a default value?

**Review by kamiranoff — APPROVED** [review]

File: packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+  getCropByRatio,
+} from "@times-components-native/image/src/utils";
+import {
+  Crop,
+  Tile,
+} from "@times-components-native/fixture-generator/src/types";
+import { Orientation } from "@times-components-native/responsive/src/context";
+
+interface Props {
+  onPress: OnArticlePress;
```
Comment (kamiranoff) [inline packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx:31]: in `packages/edition-slices/src/configured-tiles/tile-col-image-bottom/index.tsx` you have this notation:

```
interface Props {
  onPress(): OnArticlePress;
 ...
}
```

I am not too familiar with the later. Is that the same as what you have here?

**Review by kamiranoff — APPROVED** [review]

File: packages/edition-slices/src/configured-tiles/tile-col-standard/styles.tsx (hunk 14 above)
Comment (aronshamash) [reply packages/edition-slices/src/configured-tiles/tile-col-standard/styles.tsx:20]: the base config has the default value

File: packages/section/src/section.tsx (hunk 12 above)
Comment (justadamhopkins) [reply packages/section/src/section.tsx:172]: yes that was a quick fix due to MM late comments etc. Orientation was never a thing I think the transform stuff and the way we tailor variants could be improved on another iteration.

File: packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx (hunk 15 above)
Comment (justadamhopkins) [reply packages/edition-slices/src/configured-tiles/tile-col-standard/index.tsx:31]: good spot this has been fixed.

File: packages/edition-slices/src/configured-tiles/tile-col-image-bottom/index.tsx (hunk 9 above)
Comment (justadamhopkins) [reply packages/edition-slices/src/configured-tiles/tile-col-image-bottom/index.tsx:64]: good spot fixed

**Pull request merged**